### PR TITLE
B12

### PR DESCRIPTION
**Pull request by vinniefalco**
Improvements to the unit test framework, needed by NuDB

Reviewers: @mellery451 @willwray

### CODE REVIEW

File: extras/beast/unit_test/main.cpp [hunk 1]
```diff
@@ -80,18 +80,18 @@ int main(int ac, char const* av[])
...
 #endif
 
     namespace po = boost::program_options;
     po::options_description desc("Options");
     desc.add_options()
-        ("help,h",  "Produce a help message")
-        ("print,r", "Print the list of available test suites")
-        ("suites,s", po::value<string>(), "suites to run")
+       ("help,h",  "Produce a help message")
+       ("print,r", "Print the list of available test suites")
```
Comment (mellery451) [inline extras/beast/unit_test/main.cpp]: the shortcut arg for print is "r"? Is "-p" already taken?

Comment (vinniefalco) [inline extras/beast/unit_test/main.cpp]: No idea why the shortcut is `r` instead of `p`

Comment (vinniefalco) [inline extras/beast/unit_test/main.cpp]: This was a whitespace-only change

File: extras/beast/unit_test/amount.hpp [hunk 2]
```diff
@@ -23,29 +23,29 @@ class amount
...
+    operator<<(std::ostream& s, amount const& t);
 };
 
 template<class>
-amount::amount (std::size_t n, std::string const& what)
-    : n_ (n)
-    , what_ (what)
+amount::amount(std::size_t n, std::string const& what)
+    : n_(n)
+    , what_(what)
```
Comment (willwray) [inline extras/beast/unit_test/amount.hpp]: Are we switching to curly-braces in constructor initializer lists?
(except for the awkward-squad)

Comment (vinniefalco) [inline extras/beast/unit_test/amount.hpp]: I don't want to be on the cutting edge of that, as compiler support is largely untested.

**Comment by HowardHinnant** [comment]
<nit-picky>  The sub-commit message:
`Also fixes a bug where suite::log was incorrectly flushed.`
Seems a little confusing since `flush` was added, not removed.  However I don't like the double negative in this obvious alternative: `Also fixes a bug where suite::log was incorrectly not flushed.`

Maybe:
`Added flush() as needed to suite::log`
?

**Comment by vinniefalco** [comment]
@HowardHinnant Well there's much more to it than that, the entire algorithm for how the log is flushed is different (see the change calling `log` instead of `on_log`). I could go into more depth..not sure its worth it since this is a change to the unit test code. My thinking, is that from the users perspective the log was flushed incorrectly before, and now its fixed.

File: extras/beast/unit_test/amount.hpp (hunk 2 above)
Comment (HowardHinnant) [inline extras/beast/unit_test/amount.hpp]: "awkward-squad"  Going to have to put that one on a bumper sticker! :-)

Fwiw, I got hit with being too-cutting-edge with `{}` just today on my date lib: https://github.com/HowardHinnant/date/issues/78

File: extras/beast/unit_test/suite.hpp [hunk 3]
```diff
@@ -158,74 +161,103 @@ class suite
...
-    template<class F, class String>
+    template<class Condition>
     bool
-    except (F&& f, String const& reason);
+    expect(Condition const& shouldBeTrue,
+        std::string const& reason, char const* file, int line);
+    /** @} */
 
+    //
+    // DEPRECATED
```
Comment (mellery451) [inline extras/beast/unit_test/suite.hpp]: can template methods be marked `[[deprecated]]`...and would we want to do that here?

**Comment by mellery451** [comment]
👍

File: extras/beast/unit_test/suite.hpp (hunk 3 above)
Comment (vinniefalco) [inline extras/beast/unit_test/suite.hpp]: If this was an official API then yeah I would consider it but its in the extras/ directory which means its not official (but still available for use if you accept the risks).

**Comment by willwray** [comment]
👍

**Comment by nbougalis** [comment]
:+1:

**Pull request merged**